compiler/hash-typecheck/src/exhaustiveness/lower.rs

```diff
@@ -135,123 +54,319 @@ impl<'tc> LowerPatOps<'tc> {
         Self { storage }
     }
 
-    /// Take a [primitives::Pat] and convert it into [Pat], essentially lowering
-    /// the pattern.
-    pub fn lower_pat(&self, id: PatId) -> Pat {
+    /// Convert a [Pat] into a [DeconstructedPat].
+    pub(crate) fn deconstruct_pat(&self, ty: TermId, pat_id: PatId) -> DeconstructedPat {
         let reader = self.reader();
+        let pat = reader.get_pat(pat_id);
+
+        let (ctor, fields) = match pat {
+            Pat::Wild
+            | Pat::Mod(_)
+            | Pat::Binding(_)
+            | Pat::Spread(_)
+            | Pat::Access(_)
+            | Pat::Const(_) => (DeconstructedCtor::Wildcard, vec![]),
```

Comment:
How is const a wildcard?

Reply:
It has the same properties as a wild pattern in terms of exhaustiveness

Reply:
Const is just a constant member of an enum; `Color := enum(Red, Green, Blue); match c { Red => ... }` -> `Red` is `Pat::Const`.

Reply:
Yeah, so that's one of the cases that should be handled, so presumably,  for `Access`  and `Const`, it should try and get the type of the actual pattern that's passed in and create a Destructor based on that...

Does this mean that the only valid type to correspond to an access pattern would be an enum variant?

Reply:
Not necessarily, `my_mod::a::b::MyStruct(a)` is a valid pattern and is first and foremost `Pat::Access`

Reply:
yeah then it is a `Constructor` pat which is handled differently

Reply:
Yeah. Also to note that `Const` is logically equivalent to `Constructor` with zero arguments.

Reply:
Yes, it is, so I think that the only valid case for the type would be an enum, like say `None`?

Reply:
Yes. However, with unions, this essentially translates to unit structs; like `None := unit(); Some := <T> => struct(value: T); Option := <T> => type Some<T> | None`. There isn't syntax for this, but internally this will need to happen (i.e. nominal defs are either structs or units)

Reply:
The only difference between `unit()` and `struct()` is that you don't need `()` to instantiate a unit.

Reply:
I will split the two branches, but I think it will be clearer to implement this when unions replace enums internally, so I'll just leave a `todo!()`